File: solutions/ruby/isbn-verifier/3/isbn_verifier.rb
class IsbnVerifier

  VALID_ISBN10_FORMAT = /\A[[:digit:]]{9}[[:digit:]X]\z/
  DIGIT_VALUE = ->(digit) { digit == 'X' ? 10 : digit.to_i }

  def self.valid? isbn
    new(isbn).validate
  end

  private

  attr_accessor :isbn

  def initialize isbn
    self.isbn = isbn.delete '-'
  end

  def checksum
    isbn.each_char.with_index.sum { |digit, i|
      DIGIT_VALUE.(digit) * (10 - i)
    }
  end

  public

  def validate
    isbn.match?(VALID_ISBN10_FORMAT) &&
      checksum.modulo(11).zero?
  end

end

File: solutions/elixir/isbn-verifier/3/lib/isbn_verifier.ex
defmodule IsbnVerifier do

  def isbn?(isbn),
    do: isbn 
    |> String.replace("-", "") 
    |> String.to_charlist() 
    |> valid?(1, 0)

  defp valid?([], 11, checksum), do: rem(checksum, 11) == 0
  
  defp valid?([], _position, _checksum), do: false
  
  defp valid?([?X | remaining], 10, checksum),
    do: valid?(remaining, 11, checksum + 100)
    
  defp valid?([digit | remaining], position, checksum) when
    digit >= ?0 and digit <= ?9,
    do: valid?(remaining, position + 1, checksum + (digit - ?0) * position)
    
  defp valid?([_invalid | _remaining], _position, _checksum), do: false
  
end
